cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1527,4 +1531,44 @@ private void emitTimeMetrics(String namespace, String appName, String metricName
     long timeTaken = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
     metricsCollectionService.getContext(tags).gauge(metricName, timeTaken);
   }
+
+  /**
+   * Update scm metadata for the given apps.
+   * This method is to be used only in case of SCM long-running operations (like pushing
+   * multiple applications to a repository in bulk).
+   *
+   * @param namespace {@link NamespaceId} where the apps are deployed
+   * @param appsRequest {@link UpdateMultiSourceControlMetaReqeust}
+   * @throws IOException when the update operation fails
+   * @throws BadRequestException when multiple filehashes are provided for an application
+   */
+  public void updateSourceControlMeta(NamespaceId namespace, UpdateMultiSourceControlMetaReqeust appsRequest)
+      throws IOException, BadRequestException {
+    Map<ApplicationId, SourceControlMeta> updateScmMetaRequests = new HashMap<>();
+    for (UpdateSourceControlMetaRequest appRequest : appsRequest.getApps()) {
+      try {
+        ApplicationId appId = namespace.app(appRequest.getName(), appRequest.getAppVersion());
+        if (appRequest.getGitFileHash().isEmpty() || null == appRequest.getGitFileHash()) {
+          throw new BadRequestException(String.format(
+              "Empty gitFileHash for app (name = %s, version = %s).",
+              appRequest.getName(), appRequest.getAppVersion()
+          ));
+        }
+        SourceControlMeta scmMeta = new SourceControlMeta(appRequest.getGitFileHash());
+
+        if (null != updateScmMetaRequests.put(appId, scmMeta)) {
+          throw new BadRequestException(String.format(
+              "Setting multiple filehashes for an application (name = %s, version = %s) is not supported.",
+              appRequest.getName(), appRequest.getAppVersion()
+          ));
+        }
+      } catch (IllegalArgumentException | NullPointerException e) {
+        throw new BadRequestException(
+            String.format("Invalid application name (%s) or version (%s).",
```

Comment:
use the exception's message `new BadRequestException(e.getMessage(), e)`. The cause already has a more specific error message about what is wrong.

Reply:
done.